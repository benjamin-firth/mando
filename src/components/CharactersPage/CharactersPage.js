import React, { Component } from 'react';
import Character from '../Character/Character.js';
import FilmScript from '../FilmScript/FilmScript.js';

class CharactersPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      characters: props.currentMovie.characters,
    }
  }

  findCharacters = () => {
    let firstTen = []
    let i = 0;
    while (i < 10) {
      firstTen.push(this.state.characters[i]);
      i++;
    }
    return firstTen;
  }

  getFilmData = (films) => {
    let filmPromises = films.map(film => {
      return fetch(film)
        .then(res => res.json())
    })
    return Promise.all(filmPromises)
  }

  getIndividualData = (character) => {
    let homeworld = fetch(character.homeworld)
                      .then(res => res.json())

    let species = fetch(character.species[0])
                      .then(res => res.json())

    let films = this.getFilmData(character.films);

    return Promise.all([homeworld, species, films])
      .then(fetches => {
        let filmNames = fetches[2].map(movie => movie.title);
        let fullCharacterObject = {
          name: character.name,
          homeworld: fetches[0].name,
          population: fetches[0].population,
          species: fetches[1].name,
          films: filmNames
        }
        return fullCharacterObject;
      })
  }

  checkFave = (character) => {
    if (this.props.faves.find(fave => fave.name === character.name)) {
      return true;
    } else {
      return false;
    }
  }

  componentDidMount() {
    let characters = this.findCharacters();
    let promises = characters.map(character => {
      return fetch(character)
        .then(res => res.json())
        .then(data => this.getIndividualData(data))
    })
    Promise.all(promises)
      .then(data => {
        this.setState({ characters: data })
      })
  }

  render() {
    let characters = this.state.characters.map(character => {
      return <Character {...character} isFave={this.checkFave(character)} updateFave={this.props.updateFave}
      key={character.name} />
    })

    return (
      <section className="characters-container">
        <FilmScript 
          crawl={this.props.currentMovie.opening_crawl}
          filmTitle={this.props.currentMovie.title}/>
        <section className="characters">
          {characters}
        </section>  
      </section>
    )
  }
}

export default CharactersPage;
