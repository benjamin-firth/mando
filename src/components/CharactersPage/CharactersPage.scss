.characters-container {
  display: flex;
  flex-direction: row;
}

.characters {
  width: 50%;
}