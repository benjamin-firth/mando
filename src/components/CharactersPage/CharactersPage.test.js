import React from 'react';
import { shallow } from 'enzyme';
import CharactersPage from './CharactersPage.js';

describe('Characters Page', () => {
  let wrapper;
  let mockUpdateFave;
  let mockFaves;
  let mockCurrentMovie;

  beforeEach(() => {
    mockUpdateFave = jest.fn();
    mockFaves = [];
    mockCurrentMovie = {
      name: 'Johnny Tsunami',
      episode: 10,
      release: '1999-07-24',
      characters: ['Johnny', 'Tsunami'],
      opening: 'In a world, torn apart by the impending death of all humanity from Y2K, one surfer boy hopes to shed light on the important things in life'
    }
    wrapper = shallow(<CharactersPage
      faves={mockFaves}
      currentMovie={mockCurrentMovie}
      updateFave={mockUpdateFave}
    />);
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })
})
