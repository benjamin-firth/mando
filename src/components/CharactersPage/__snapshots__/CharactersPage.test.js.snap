// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Characters Page should match the snapshot 1`] = `
<section
  className="main"
>
  <p />
  <Character
    0="J"
    1="o"
    2="h"
    3="n"
    4="n"
    5="y"
    isFave={false}
    updateFave={[MockFunction]}
  />
  <Character
    0="T"
    1="s"
    2="u"
    3="n"
    4="a"
    5="m"
    6="i"
    isFave={false}
    updateFave={[MockFunction]}
  />
</section>
`;
